### PR TITLE
[MPLUGIN-372] Descriptor mojo fails if super class is provided scope

### PR DESCRIPTION
**Pull request by cstamas**
https://issues.apache.org/jira/browse/MPLUGIN-372

Changes:
- Bug: adjust descriptor mojo resolution scope to include provided scope as well
- drop use of deprecated `project.getRuntimeDependencies()`
- Sanitized generator and report mojos (injects both project and project.artifacts, just one is completely enough)
- GeneratorUtils using Artifacts not Dependencies (as they were not needed)
- JavaAnnotationsMojoDescriptorExtractor error reporting: report failed class at least

### CODE REVIEW

**Comment by cstamas** [comment]
This change should be addressed, as it is crucial to put plugins maven deps to provided.

**Comment by rfscholte** [comment]
By just looking at the code I don't see weird things. This ticket confirms my feeling that there was something wrong, but I can't see the impact yet. I guess you can see only the difference if MNG-7097 is ready.

**Review by michael-o — COMMENTED** [review]
I really don't have an opion on this. Robert is the expect here.

**Comment by cstamas** [comment]
@rfscholte ping.

Some explanation: MPLUGIN-372 (this PR) and MNG-7097 are not about same thing:
- this PR fixes an issue about plugin tooling that currently prevent us using `provided` scope for maven artifacts. Is needed to implement MPLUGIN-370. So is for "future" (yet to be released) plugins.
- OTOH, MNG-7097 is for "everything else" (and old plugins, not having MPLUGIN-370 implemented), but MNG-7097 needs more changes to allow cases where "all is needed", see MNG-5783... am thinking to extend Mojo annotation and make Mojos like these "ask" for full (MNG-7097 disabled) resolution?

**Comment by cstamas** [comment]
Or it could be other way around: add a Mojo annotation that plugin is "happy" with MNG-7097 enabled? This seems a bit counter-productive, as IMHO 99% of plugins are happy with it, only some exceptions like cobertura (or other crafting CP and spawning JVM may suffer from it).

**Comment by rfscholte** [comment]
To me cobertura shouldn't block us for implementing this improvement.

File: maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/GeneratorUtils.java [hunk 1]
```diff
@@ -120,14 +121,14 @@ public static void element( XMLWriter w, String name, String value, boolean asTe
     }
     
     /**
-     * @param dependencies not null list of <code>Dependency</code>
+     * @param dependencies not null collection of <code>Artifact</code>
      * @return list of component dependencies
      */
-    public static List<ComponentDependency> toComponentDependencies( List<Dependency> dependencies )
+    public static List<ComponentDependency> toComponentDependencies( Collection<Artifact> dependencies )
```
Comment (rfscholte) [inline maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/GeneratorUtils.java:127]: rename parameter to `artifacts`

File: maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/GeneratorUtils.java [hunk 2]
```diff
@@ -120,14 +121,14 @@ public static void element( XMLWriter w, String name, String value, boolean asTe
...
     {
         List<ComponentDependency> componentDeps = new LinkedList<>();
 
-        for ( Dependency dependency : dependencies )
+        for ( Artifact dependency : dependencies )
```
Comment (rfscholte) [inline maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/GeneratorUtils.java:131]: rename variable to `artifact`

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/PluginReport.java [hunk 3]
```diff
@@ -316,13 +307,13 @@ private PluginDescriptor extractPluginDescriptor()
...
         {
-            List<ComponentDependency> deps = GeneratorUtils.toComponentDependencies( project.getRuntimeDependencies() );
+            List<ComponentDependency> deps = GeneratorUtils.toComponentDependencies( project.getArtifacts() );
             pluginDescriptor.setDependencies( deps );
 
             PluginToolsRequest request = new DefaultPluginToolsRequest( project, pluginDescriptor );
             request.setEncoding( encoding );
             request.setSkipErrorNoDescriptorsFound( true );
-            request.setDependencies( dependencies );
+            request.setDependencies( new LinkedHashSet<>( project.getArtifacts() ) );
```
Comment (rfscholte) [inline maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/PluginReport.java:316]: Are we adding artifacts or dependencies?

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java [hunk 4]
```diff
@@ -313,7 +305,7 @@ static String getDefaultGoalPrefix( MavenProject project )
         Set<Artifact> filteredDependencies;
         if ( mojoDependencies == null )
         {
-            filteredDependencies = dependencies;
+            filteredDependencies = new LinkedHashSet<>( project.getArtifacts() );
```
Comment (rfscholte) [inline maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java:308]: dependencies or artifacts?

**Review by rfscholte — CHANGES_REQUESTED** [review]
In general we need to be clear if we are talking about dependencies or artifacts.
http://maven.apache.org/shared/maven-artifact-transfer/comparison.html tries to explain the difference.

File: maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/GeneratorUtils.java (hunk 1 above)
Comment (cstamas) [reply maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/GeneratorUtils.java]: fixed

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/PluginReport.java (hunk 3 above)
Comment (cstamas) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/PluginReport.java:316]: fixed

File: maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/GeneratorUtils.java (hunk 2 above)
Comment (cstamas) [reply maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/GeneratorUtils.java]: fixed

**Comment by cstamas** [comment]
@rfscholte for the rest of your comments: well, as it can be seen, these are all uses of `org.apache.maven.project.MavenProject#getArtifacts` method, which in turn has JavaDoc saying "All dependencies that this project has...". Also, old (removed) code was calling it "dependencies", so I did not change that. The util class is aligned though.

**Comment by cstamas** [comment]
ping

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java (hunk 4 above)
Comment (rfscholte) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java:303]: Renaming `filteredDependencies` to `filteredArtifacts` should be sufficient, right?

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/PluginReport.java (hunk 3 above)
Comment (rfscholte) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/PluginReport.java:316]: There's nothing we can do here anymore. Could be a different ticket to fix this inconsistency.

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java (hunk 4 above)
Comment (cstamas) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java]: done

**Review by rfscholte — APPROVED** [review]

**Comment by cstamas** [comment]
merged as 8f354439765afdbf6ce166d299791117576e90f2

**Pull request closed without merging**